File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        int prefixProduct = 1;
        for(int i = 0 ; i < n; i++) {
            result[i] = prefixProduct;
            prefixProduct *= nums[i];
        }
        int suffixProduct = 1;
        for(int i = n-1; i >= 0; i--) {
            result[i] *= suffixProduct;
            suffixProduct *= nums[i];
        }
        
        return result;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        answer = [0] * len(nums)
        answer[0] = 1
        for i in range(1,len(nums)):
            answer[i] = answer[i-1] * nums[i-1]
        
        R =1
        for i in range(len(nums)-1,-1,-1):
            answer[i] = answer[i] * R
            R *= nums[i]
        
        return answer
        